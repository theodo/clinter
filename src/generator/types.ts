import { Linter } from "eslint";
import { ProjectInfoObject } from "types";

export type ESLintGenerator = (userAnswers: ProjectInfoObject) => (config: Linter.Config) => Linter.Config;

export type ESLintDependencyGenerator = (userAnswers: ProjectInfoObject) => (deps: string[]) => string[];

export interface ESLintOverrider {
  files: string[];
}
