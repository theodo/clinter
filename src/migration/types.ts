export interface MigrationResults {
  fixableErrorCount: number;
  errorCount: number;
}
