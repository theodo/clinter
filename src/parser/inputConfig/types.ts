import { AnswerObject, ModeAnswerObject } from "types";

export interface InputConfig {
  generatorConfig: AnswerObject;
  modeConfig: ModeAnswerObject;
}
