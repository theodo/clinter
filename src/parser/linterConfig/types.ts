import { Options } from "prettier";
import { Linter } from "eslint";

export enum FileExtension {
  JSON = "json",
  YAML = "yml",
  JS = "js",
  NONE = "none",
}

export interface ConfigFileObject {
  name: string;
  extension: FileExtension;
  attribute?: string;
}

export type LinterConfigs = Options | Linter.Config;

export interface ConfigContainer<T extends LinterConfigs> {
  config: T;
  file: ConfigFileObject;
}
