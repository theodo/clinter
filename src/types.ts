export enum FormatterInfo {
  None = "None",
  Prettier = "Prettier",
}

export interface FormatterInfoObject {
  formatter: FormatterInfo;
}

export enum FrontFrameworkInfo {
  None = "None",
  Vue = "Vue",
  React = "React",
}

export interface FrontFrameworkInfoObject {
  frontFramework: FrontFrameworkInfo;
}

export enum TypescriptInfo {
  None = "None",
  WithTypeChecking = "With type checking",
  NoTypeChecking = "Without type checking",
}

export interface TypescriptInfoObject {
  typescript: TypescriptInfo;
}

export enum EnvInfo {
  Browser = "Browser",
  Node = "Node",
}

export interface EnvInfoObject {
  env: EnvInfo[];
}

export enum TestFrameworkInfo {
  None = "None",
  Jest = "Jest",
}

export interface TestFrameworkInfoObject {
  test: TestFrameworkInfo;
}

export type ProjectInfoObject = FormatterInfoObject &
  FrontFrameworkInfoObject &
  TypescriptInfoObject &
  EnvInfoObject &
  TestFrameworkInfoObject;

export enum ClinterModeInfo {
  Generator = "Generator Mode",
  Upgrade = "Upgrade Mode",
}

export interface ClinterModeInfoObject {
  mode: ClinterModeInfo;
}

export interface MigrationModeInfoObject {
  migration: boolean;
}

export interface DependenciesConfigObject {
  upgradeDependencies: boolean;
}

export interface ClinterSettings {
  generatorConfig: ProjectInfoObject;
  modeConfig: ClinterModeInfoObject;
  migrationModeConfig: MigrationModeInfoObject;
  dependenciesConfig: DependenciesConfigObject;
}
