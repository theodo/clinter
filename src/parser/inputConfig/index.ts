export { parseInputConfigFile } from "./parseInputConfigFile";
export { parseInputConfigQuestions } from "./parseInputConfigQuestions";
