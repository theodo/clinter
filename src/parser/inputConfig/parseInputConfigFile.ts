import signale from "signale";
import { existsSync, readFileSync } from "fs";

import { InputConfig } from "parser/inputConfig/types";

export function parseInputConfigFile(inputFileName: string): InputConfig {
  if (!existsSync(inputFileName)) {
    signale.error(`Cannot find InputFile ${inputFileName}`);
    process.exit(1);
  }

  try {
    return JSON.parse(readFileSync(inputFileName, "utf-8")) as InputConfig;
  } catch (error) {
    signale.error("Error parsing InputFile configuration");
    process.exit(1);
  }
}
