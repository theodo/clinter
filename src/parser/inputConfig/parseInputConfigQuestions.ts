import signale from "signale";

import { promptGeneratorUserQuestions, promptModeUserQuestions } from "parser/userQuestions";
import { InputConfig } from "parser/inputConfig/types";

export async function parseInputConfigQuestions(): Promise<InputConfig> {
  const modeConfig = await promptModeUserQuestions();

  const generatorConfig = await promptGeneratorUserQuestions();
  signale.info("Retrieving answers ...");

  return {
    modeConfig,
    generatorConfig,
  };
}
