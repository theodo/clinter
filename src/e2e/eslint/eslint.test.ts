import { makeTestService } from "e2e/helpers";
import { InputConfig } from "parser/inputConfig/types";
import { FormatterAnswer, FrontFrameworkAnswer, ModeAnswer, TestFrameworkAnswer, TypescriptAnswer } from "types";

describe("ESLint configuration", () => {
  const testService = makeTestService();

  it("should return the right eslint configuration for the right input", async () => {
    const inputConfig: InputConfig = {
      modeConfig: {
        mode: ModeAnswer.Generator,
      },
      generatorConfig: {
        typescript: TypescriptAnswer.None,
        env: [],
        formatter: FormatterAnswer.Prettier,
        frontFramework: FrontFrameworkAnswer.None,
        test: TestFrameworkAnswer.Jest,
      },
    };

    testService.loadInputConfig(inputConfig);
    await testService.runClinter();
    const outputConfig = testService.getOutputConfig();

    expect(outputConfig).toMatchSnapshot("eslint");
  });
});
