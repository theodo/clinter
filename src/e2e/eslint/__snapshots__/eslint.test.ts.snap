// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ESLint configuration should return the right eslint configuration for the right input: eslint 1`] = `
"{
  \\"root\\": true,
  \\"env\\": {
    \\"jest\\": true
  },
  \\"parserOptions\\": {},
  \\"extends\\": [
    \\"eslint:recommended\\",
    \\"plugin:prettier/recommended\\"
  ],
  \\"plugins\\": [
    \\"import\\"
  ],
  \\"rules\\": {
    \\"curly\\": [
      \\"error\\",
      \\"all\\"
    ],
    \\"eqeqeq\\": [
      \\"error\\",
      \\"smart\\"
    ],
    \\"complexity\\": [
      \\"error\\",
      8
    ],
    \\"import/no-extraneous-dependencies\\": [
      \\"error\\",
      {
        \\"devDependencies\\": true,
        \\"optionalDependencies\\": false,
        \\"peerDependencies\\": false
      }
    ],
    \\"no-shadow\\": [
      \\"error\\",
      {
        \\"hoist\\": \\"all\\"
      }
    ],
    \\"prefer-const\\": \\"error\\",
    \\"import/order\\": [
      \\"error\\",
      {
        \\"groups\\": [
          [
            \\"external\\",
            \\"builtin\\"
          ],
          \\"internal\\",
          [
            \\"parent\\",
            \\"sibling\\",
            \\"index\\"
          ]
        ]
      }
    ],
    \\"sort-imports\\": [
      \\"error\\",
      {
        \\"ignoreCase\\": true,
        \\"ignoreDeclarationSort\\": true,
        \\"ignoreMemberSort\\": false,
        \\"memberSyntaxSortOrder\\": [
          \\"none\\",
          \\"all\\",
          \\"multiple\\",
          \\"single\\"
        ]
      }
    ]
  }
}"
`;
